include/internal/refcount.h

```diff
@@ -36,21 +51,51 @@ static inline int CRYPTO_UP_REF(CRYPTO_REF_COUNT *refcnt, int *ret)
     return 1;
 }
 
-/*
- * Changes to shared structure other than reference counter have to be
- * serialized. And any kind of serialization implies a release fence. This
- * means that by the time reference counter is decremented all other
- * changes are visible on all processors. Hence decrement itself can be
- * relaxed. In case it hits zero, object will be destructed. Since it's
- * last use of the object, destructor programmer might reason that access
- * to mutable members doesn't have to be serialized anymore, which would
- * otherwise imply an acquire fence. Hence conditional acquire fence...
+/**
+ * @brief Atomically decrements the reference count and stores the result.
+ *
+ * This function atomically decreases the reference count of a CRYPTO_REF_COUNT
+ * object. It ensures memory ordering constraints based on the architecture.
+ *
+ * @param refcnt A pointer to the CRYPTO_REF_COUNT object to be decremented.
+ * @param ret A pointer to an integer where the decremented result is stored.
+ *
+ * @return Always returns 1.
+ *
+ * Architecture-specific memory ordering is applied:
+ * - On ARM and AArch64, it uses __ATOMIC_ACQ_REL.
+ * - On other architectures, it uses __ATOMIC_RELAXED and ThreadSanitizer
+ *   acquire/release barriers.
+ *
+ * If the resulting reference count is 0, a memory fence is issued to ensure
+ * proper ordering.
+ *
+ * NOTE: Memory ordering is arch dependent here.  On strongly ordered systems
+ * a relaxed ordering is acceptable, because on such systems (like x86), the
+ * hardware ensures that atomic writes are visible to other cpus immediately.
+ *
+ * However, on weakly ordered systems, and atomic read/write is only atomic with
+ * respect to the local cpu.  To ensure that other cpus see a modified value, we
+ * need to use an atomic aquire/release ordering here, which emits the
+ * appropriate ldxar and stlxr instructions to ensure load/acquire, and
+ * store/release semantics.
+ *
+ * Note further that on x86 platforms, the subsequent atomic fence instruction
+ * isn't interpreted properly by thread sanitizers, and so, with relaxed
+ * ordering, tsan still potentially complains about multithreaded read/write
+ * races.  To avoid incurring the stricter semantic constraints there, we keep
+ * the relaxed ordering (sine the hardware enforces it anyway), and instead
+ * inform tsan that its ok with the __tsan_acquire and __tsan_release notations
  */
 static inline int CRYPTO_DOWN_REF(CRYPTO_REF_COUNT *refcnt, int *ret)
 {
-    *ret = atomic_fetch_sub_explicit(&refcnt->val, 1, memory_order_relaxed) - 1;
-    if (*ret == 0)
-        atomic_thread_fence(memory_order_acquire);
+#    if defined (__arm__) || defined (__aarch64)
+    *ret = atomic_fetch_sub_explicit(&refcnt->val, 1, __ATOMIC_ACQ_REL) - 1;
```

Comment:
I strongly believe that we should always specify acquire/release here and forget about the fences.

Furthermore, are we certain that all architectures we support apart from ARM and x86 provide the very strong x86 memory guarantees?  If we must optimise this, we should be checking for x86 and defaulting to the ARM code here which will work universally.

Have we measured the performance difference between acquire/release on x86 and the fenced version? Is there a performance difference in favour of the latter? Without such a measurement indicating a performance gain, this is premature optimisation.

Reply:
In answer to your question, no, I am absolutely not sure.  I'm fairly familiar with arm and x86, but I have no idea about mips/ia64/whatever arch nonstop uses/etc.

I've not measured performance differences but I expect there will be some on x86 between relaxed ordering and aquire/release ordering, as the latter will introduce cache invalidations, though addressing x86 specifically here seem like a good idea to me, just a bit coutner-intuitive as most of the common code somewhat presumes x86 by default.

Reply:
FWIW, did some quick runs of CRYPTO_DOWN_REF using our test harness

In its current state (using a relaxed fetch_sub + a conditional fence) we average 0.10325 usec/down ref call

Modifying this PR to always do an acquire/release ordering without a fence yields 0.105173 usecs/down ref call

Both on intel hardware running 10 threads hitting the same reference counter.

Thats a 1.8% difference, which I think would be largely noise on any real system.  Seems like it makes sense to just  do acquire/release unilaterally and save the per arch conditioning.

Reply:
I kind of suspected that the x86's strong memory ordering would effectively do something equivalent to acq/rel regardless of the specified ordering.

Reply:
@nhorman What are you benchmarking?
on x86 both code compiles to same asm

So this difference is difference between different runs of your benchmark

Reply:
Even more, on x86, **rmw operations** always look same, no matter what memory order, you can compare relaxed and seq_cst in any case.

Reply:
Yeah, so x86 perf isn't relevant. What is relevant is ARM but not with 10 threads on some regular machine but 100 or more threads on some multi-core multi-cpu high end machine. We already have problems with perf on such systems (albeit I do not know which architecture).

Reply:
@t8m On arm relaxed will compiles to incorrect asm

Reply:
We could compare the release + acquire fence with rel_acq there.

Also it would be interesting to know if the test from #25664 causes real memory leaks or any other errors on ARM with the current supposedly incorrect code.

Reply:
> We could compare the release + acquire fence with rel_acq there.

Yes, it's interesting!